proposals/semi-auto-properties.md

```diff
@@ -1,14 +1,325 @@
+{
+    public C()
+    {
+        P1 = 1; // Assigns P1's backing field directly
+        P2 = 2; // Assigns P2's backing field directly
+        P3 = 3; // Calls P3's setter
+        P4 = 4; // Calls P4's setter
+    }
+
+    public int P1 => field;
+    public int P2 { get => field; }
+    public int P4 { get => field; set => field = value; }
+    public int P3 { get => field; set; }
+}
+```
+
+### Definite assignment in structs
+
+Even though they can't be referenced in the constructor, backing fields denoted by the `field` keyword are subject to default-initialization and disabled-by-default warnings under the same conditions as any other struct fields ([LDM decision 1](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-03-02.md#property-assignment-in-structs), [LDM decision 2](https://github.com/dotnet/csharplang/blob/main/meetings/2022/LDM-2022-05-02.md#definite-assignment-of-manually-implemented-setters)).
+
+For example (these diagnostics are silent by default):
+
+```cs
+public struct S
+{
+    public S()
+    {
+        // CS9020 The 'this' object is read before all of its fields have been assigned, causing preceding implicit
+        // assignments of 'default' to non-explicitly assigned fields.
+        _ = P1;
+    }
+
+    public int P1 { get => field; }
+}
+```
+
+```cs
+public struct S
+{
+    public S()
+    {
+        // CS9020 The 'this' object is read before all of its fields have been assigned, causing preceding implicit
+        // assignments of 'default' to non-explicitly assigned fields.
+        P2 = 5;
+    }
+
+    public int P1 { get => field; }
+}
+```
+
+### Nullability
+
+When `{ get; }` is written as `{ get => field; }`, or `{ get; set; }` is written as `{ get => field; set => field = value; }`, a similar warning should be produced when a non-nullable property is not initialized:
+
+```cs
+class C
+{
+    // ⚠️ CS8618: Non-nullable property 'P' must contain a
+    // non-null value when exiting constructor.
+    public string P { get => field; set => field = value; }
+}
+```
+
+No warning should be produced if the property is initialized to a non-null value via constructor assignment or property initializer:
+
+```cs
+class C
+{
+    public C() { P = ""; }
+
+    public string P { get => field; set => field = value; }
+}
+```
+
+```cs
+class C
+{
+    public string P { get => field; set => field = value; } = "";
+}
+```
+
+#### Open question: nullability of `field`
+
+In the same vein as how `var` infers as nullable for reference types, the `field` type is the nullable type of the property whenever the property's type is not a value type. Otherwise, `field ??` would appear to be followed by dead code, and it avoids producing a misleading warning in the following example:
+
+```cs
+public string AmbientValue
+{
+    get => field ?? parent.AmbientValue;
+    set
+    {
+        if (value == parent.AmbientValue)
+            field = null; // No warning here. Resume following the parent's value.
```

Comment:
This feels wrong to me. Feel like this should be a warning and users complaining about that warning would be evidence that we should expand the feature in the future to allow explicit declaration of the field type.

Reply:
Great consideration, being discussed in WG.